### DISCUSSION TITLE
How to Reduce Buffering Time On HLS?

### DISCUSSION THREAD

**Discussion by aggo15** [bluenviron/mediamtx#679] [Questions]
Thanks @aler9 for this wonderful piece of software. I got one question regarding the HLS stream as mentioned in subject.

My setup: 
1. Raspberry Pi Zero W with official camera board.
2. Running "aler9/rtsp-simple-server" docker image on the said device.
3. Using raspivid to redirect the video feed to ffmpeg, ffmpeg publish the video stream to docker container using rtmp protocol.
4. The setup are connected to a local area network (wifi 2.4GHz), not streaming over the internet.

Observation:
1. No issue capturing the video stream using OBS studio, running on rtmp protocol. Video playback is acceptable. (15fps)
**2. When accessing the video stream via HLS (http://ip-pi-zero:8888/mystream), video able to play for n amount of second and it start buffering for few seconds before continue playing. The buffering keep happening after n amount of second causing the video playback not smooth.**

I've tried to change the value for hlsSegmentCount and hlsSegmentDuration in configuration, but seems like it didn't improve the buffering time. If I set hlsSegmentDuration to 10 seconds, it seems like it only start loading the next chuck after it hit 10 seconds mark. I expect it should start loading the next chunk before it hit 10 seconds mark. 

I also tried to stream video over ethernet by connecting the Pi Zero to USB ethernet adapter, the issue didn't go away. I guess this is not related to network issue.

Is there any configuration I can change to improve the video playback over HLS? Any help is appreciated.

Thank you.

**Comment by aler9** [comment]
hello, you can

1. try tuning the interval between key frames in rpicamsrc:

   ```
   gst-launch-1.0 rpicamsrc preview=false bitrate=2000000 keyframe-interval=100
   ```

   Greater interval corresponds to more robust buffering, lower to a lowest latency

2. try decreasing the bitrate:

   ```
   gst-launch-1.0 rpicamsrc preview=false bitrate=500000 keyframe-interval=50
   ```

**Reply by aggo15** [reply]
Hi @aler9, 

Thanks for getting back to me. Are those gstreamer commands? Is there a equivalent way to do adjust the key frame interval or decrease the bit rate with raspivid command instead?

Thank you.

**Reply by aler9** [reply]
raspivid should support the same exact parameters, but anyway, raspivid is not the supported way to fed a Raspberry Camera into the server. Please take a look at the README.

**Reply by aggo15** [reply]
Hi @aler9,

I've tried the parameters you suggested on raspivid, seems like no improvement at all. I suspect that Pi zero is a little under power to run both the server and raspivid at the same time. 

What I did is let Pi zero focusing on just pushing video stream to the server sitting inside another Pi. With this configuration, I no longer see HLS video stream stuttering anymore, even if I push video stream via wifi.

So I guess we can close this case? Is a shame I can't run everything in one Pi but there is nothing we can do about it. =(

**Comment by Curid** [comment]
FFmpeg can read the camera and output HLS. Why use raspivid and rtsp-simple-server?

Anyway, its probably buffering because rtsp-simple-server uses a really slow mpegts muxer.

**Reply by aggo15** [reply]
Initially the idea was quickly setup a working gear for streaming at work, as a poc and see how well it run out of the box. I'm not familiar with gstreamer/ffmpeg so I opt to use something else that is already available, easy to use and not require to build something out from source.

Currently I'm using prebuilt docker image from balenalib/raspberrypi3 with raspivid available inside the image + rtsp-simple-server docker image, manage it via docker compose. It is easier to use this way in my opinion.

**Reply by aler9** [reply]
> Anyway, its probably buffering because rtsp-simple-server uses a really slow mpegts muxer.

It's not slow at all, but anyway both the raspberry camera support and the HLS support can be improved. It would be great to support the Raspberry Camera without raspicamsrc, and it would be great to support low latency-HLS. Actually i'm working on the HLS side.

> Currently I'm using prebuilt docker image from balenalib/raspberrypi3 with raspivid available inside the image + rtsp-simple-server docker image, manage it via docker compose. It is easier to use this way in my opinion.

Great idea

**Reply by Curid** [reply]
> It's not slow at all

go-astits is about 10x slower compared to FFmpeg.

**Reply by aler9** [reply]
@Curid performance and latency are completely unrelated in this case. Performance, which you are referring to, affects CPU usage, not latency.

**Reply by Curid** [reply]
@aler9 If the buffering is caused by the server getting overloaded and dropping packets then it is.

"video able to play for n amount of second and it start buffering"

**Reply by aler9** [reply]
@Curid you're not helping at all, furthermore you're spreading undocumented informations that unfortunately a lot of people are gonna see when they search on Google about the topic. I've never seen the server consuming an entire CPU, even with 500 sources. Regarding the comparison between ffmpeg and a Golang library, it has no basis.

For the moment i'm locking this thread, next time i'm reporting your comments.
